blockchain/chainio.go

```diff
@@ -1271,6 +1290,60 @@ func dbFetchBlockByNode(dbTx database.Tx, node *blockNode) (*btcutil.Block, erro
 	return block, nil
 }
 
+// dbStoreBlock stores the provided block in the database. The block header is
+// written to the block index bucket and full block data is written to ffldb.
+func dbStoreBlockHeader(dbTx database.Tx, blockHeader *wire.BlockHeader, height uint32) error {
+	// Serialize block data to be stored. This is just the serialized header.
+	w := bytes.NewBuffer(make([]byte, 0, blockHdrSize))
+	err := blockHeader.Serialize(w)
+	if err != nil {
+		return err
+	}
+	value := w.Bytes()
+
+	// Write block header data to block index bucket.
+	blockHash := blockHeader.BlockHash()
+	blockIndexBucket := dbTx.Metadata().Bucket(blockIndexBucketName)
+	key := blockIndexKey(&blockHash, height)
+	return blockIndexBucket.Put(key, value)
+}
+
+// dbStoreBlock stores the provided block in the database. The block header is
+// written to the block index bucket and full block data is written to ffldb.
+func dbStoreBlock(dbTx database.Tx, block *btcutil.Block) error {
+	if block.Height() == btcutil.BlockHeightUnknown {
+		return fmt.Errorf("cannot store block %s with unknown height",
+			block.Hash())
+	}
+
+	// First store block header in the block index bucket.
+	err := dbStoreBlockHeader(dbTx, &block.MsgBlock().Header,
```

Comment:
Yeah, no reason not to do the check here, but I skipped it under the assumption that it would only be called if the block does not already exist in the DB (which is true currently AFAICT). Will add.

Reply:
I realized I hadn't answered this yet.  I agree with you that the assumption is true currently, however, once everything is completely headers-first, won't this function will be called with the full block after the header has already previously been stored during the initial header download and store pass?

Reply:
Can we leave this as is for this PR? https://github.com/btcsuite/btcd/pull/1066 has to change the behavior of this method anyway so that `dbStoreBlock` does not call `dbStoreBlockHeader`.

Reply:
I'm fine with leaving it as is for now.  However, did you want to put this PR in separately or roll it up into #1066 for a single migration path?  This is ready to be merged as far as the code it does itself, however, if we merge it now, it needs the extra migration code.